### PR TITLE
Break down socket connect into smaller pieces

### PR DESCRIPTION
**Pull request by zhangxin511**
Socket connect flow is very hard to follow and read. This PR do followings:
1. Introduced `ConnectDocumentStage` enum to indicate different of stages, which will be used for reliability reporting.
2. Break the long connect method into smaller functions, so that code is easier to scan/read.
3. Moved version check step to the first step instead of last few steps. 

Need to load diff for the large diff view

### CODE REVIEW

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 1]
```diff
@@ -109,6 +116,22 @@ function selectProtocolVersion(connectVersions: string[]): string | undefined {
...
+}
+
+function checkVersion(versions: string[]): [string[], string] {
+	// Iterate over the version ranges provided by the client and select the best one that works
+	const connectVersions = versions ? versions : ["^0.1.0"];
+	const version = selectProtocolVersion(connectVersions);
+	if (!version) {
+		throw new NetworkError(
+			400,
+			`Unsupported client protocol. Server: ${protocolVersions}. Client: ${JSON.stringify(
```
Comment (alexvy86) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:129]: `protocolVersions` and `connectVersions` are both string arrays so either both of them need `JSON.stringify()`, or neither does, I think?

**Review by alexvy86 — COMMENTED** [review]
I didn't go through the bulk of the changes but added a few comments.

Also, we don't have a formal process for this yet, but some conversations we've had with Loop make me really want to maintain a changelog for relevant changes in server code, so we can eventually produce release notes for server releases. "An invalid JWT token will now return a 401 error instead of a 500" feels like a great candidate for that. (And if anything in the file that I didn't fully look at has similar end-user visible changes, those too).

In the client release group we use [changesets](https://github.com/changesets/changesets) for this purpose (cc @tylerbutler), I think eventually we should end up with the same tooling in both release groups. But if we don't want to introduce them in server just yet, I'd at least ask that we create a `CHANGELOG.md` file under `server/routerlicious` and start putting them there.

File: server/routerlicious/packages/services-utils/src/auth.ts [hunk 2]
```diff
@@ -318,7 +318,13 @@ export function validateTokenScopeClaims(expectedScopes: string): RequestHandler
 			);
 		}
 
-		const claims = decode(token) as ITokenClaims;
+		let claims: ITokenClaims;
+		try {
+			claims = decode(token) as ITokenClaims;
+		} catch {
+			throw new NetworkError(401, "Invalid token");
```
[the remark this answers is not in the document]
Comment (znewton) [reply server/routerlicious/packages/services-utils/src/auth.ts:325]: Correct. This is a middleware. If you throw here it will have to be caught by a global error catcher which has less functionality than directly responding with the network error.

[the remark this answers is not in the document]
Comment (znewton) [reply server/routerlicious/packages/services-utils/src/auth.ts:325]: Also I don't understand why this is being changed in this PR. It has nothing to do with Socket connection

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 3]
```diff
@@ -65,15 +67,20 @@ const getMessageMetadata = (documentId: string, tenantId: string) => ({
...
-	Lumberjack.error(errorMessage, getLumberBaseProperties(documentId, tenantId));
-	throw new NetworkError(500, "Failed to connect client to document.");
+	error: any,
+): NetworkError => {
+	const errMsgWithPrefix = `Connect Cerver Error - ${errorMessage}`;
+	logger.error(errMsgWithPrefix, { messageMetaData: getMessageMetadata(documentId, tenantId) });
+	Lumberjack.error(errMsgWithPrefix, getLumberBaseProperties(documentId, tenantId), error);
+	return new NetworkError(
+		500,
+		"Failed to connect client to document. Check correlation Id for details.",
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:82]: That is more like for me, but I will add it here

File: server/routerlicious/packages/services-utils/src/auth.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/services-utils/src/auth.ts]: I started with just category some errors of connect, and to distinguish eac. I noticed this 401 error somehow becomes 500 on server side, so added these in this PR: 
image


And when I read code in socket connect, it make me feel almost impossible to read/follow the code, so I decided to do the refactor, which lead to the majority change.

Given the auth code change is minimum, I decided to leave them there, so I don't have to create another PR

**Comment by zhangxin511** [comment]
> I didn't go through the bulk of the changes but added a few comments.
> 
> Also, we don't have a formal process for this yet, but some conversations we've had with Loop make me really want to maintain a changelog for relevant changes in server code, so we can eventually produce release notes for server releases. "An invalid JWT token will now return a 401 error instead of a 500" feels like a great candidate for that. (And if anything in the file that I didn't fully look at has similar end-user visible changes, those too).
> 
> In the client release group we use [changesets](https://github.com/changesets/changesets) for this purpose (cc @tylerbutler), I think eventually we should end up with the same tooling in both release groups. But if we don't want to introduce them in server just yet, I'd at least ask that we create a `CHANGELOG.md` file under `server/routerlicious` and start putting them there.

Put these suggestions in BREAKING.md, given I think error code from 500 to 401 for bad jwt is still a breaking change

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: Added the correlationId

File: server/routerlicious/packages/lambdas/src/alfred/index.ts (hunk 1 above)
Comment (zhangxin511) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: Nice catch, this was from old code, all I did was just extract to a new method. Addressed

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 4]
```diff
@@ -206,6 +229,26 @@ function checkThrottleAndUsage(
...
+	TokenExpirySet = "TokenExpirySet",
+	MessageClientConnected = "MessageClientConnected",
+	SocketTrackerAppended = "SocketTrackerAppended",
+	SignalListenerSetUp = "SignalListenerSetUp",
+	JoinOpEmitted = "JoinOpEmitted",
+}
+
+const hasWriteAccess = (scopes: string[]) => canWrite(scopes) || canSummarize(scopes);
+const isWriter = (scopes: string[], mode: ConnectionMode) =>
+	hasWriteAccess(scopes) ? mode === "write" : false;
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: Good find. This was from old code, all I did was just extract to a new method. Addressed

**Comment by tylerbutler** [comment]
> In the client release group we use [changesets](https://github.com/changesets/changesets) for this purpose (cc @tylerbutler), I think eventually we should end up with the same tooling in both release groups.

@scottn12 started looking at changesets for server changesets just this week. We have some work items on the schedule.

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 5]
```diff
@@ -109,6 +116,22 @@ function selectProtocolVersion(connectVersions: string[]): string | undefined {
 			}
 		}
 	}
+	return undefined;
+}
+
+function checkVersion(versions: string[]): [string[], string] {
+	// Iterate over the version ranges provided by the client and select the best one that works
+	const connectVersions = versions ? versions : ["^0.1.0"];
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: This was old code, I have it addressed

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 6]
```diff
@@ -65,15 +67,20 @@ const getMessageMetadata = (documentId: string, tenantId: string) => ({
...
 	errorMessage: string,
 	documentId: string,
 	tenantId: string,
-) => {
-	logger.error(errorMessage, { messageMetaData: getMessageMetadata(documentId, tenantId) });
-	Lumberjack.error(errorMessage, getLumberBaseProperties(documentId, tenantId));
-	throw new NetworkError(500, "Failed to connect client to document.");
+	error: any,
+): NetworkError => {
+	const errMsgWithPrefix = `Connect Cerver Error - ${errorMessage}`;
```
[the remark this answers is not in the document]
Comment (zhangxin511) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: Sorry for the typo, addressed

**Comment by sharmakhushboo** [comment]
@zhangxin511 , @alteut is working on moving the socket logic from alfred to a new service called Nexus. 

I think it might be better to make these changes in Neuxs once Alejandro's pr is merged?
